tmt/steps/execute/internal.py

```diff
@@ -221,53 +229,129 @@ def check(self, test):
         guest.push(self.workdir, options=DEFAULT_RSYNC_OPTIONS + ["-p"])
         return template, setup, teardown
 
+    def _setup_rstrnt_report_result_script(self, guest):
+        """ Set up rstrnt-report-result script on the guest. """
+        script_path = os.path.dirname(
+            __file__) + REPORT_RESULT_RESULT_SITE_PATH
+        with open(script_path) as f:
+            contents = f.read()
+            f.close()
+        template = self._setup_tmt_restraint_file(
+            REPORT_RESULT_TEMPLATE_NAME,
+            RSTRNT_REPORT_RESULT_SCRIPT.format(contents))
+        setup = \
+            self._setup_tmt_restraint_file(REPORT_RESULT_SETUP_NAME,
+                                           TMT_RESTRAINT_SETUP_SCRIPT,
+                                           RSTRNT_REPORT_RESULT_SCRIPT_PATHS)
+        teardown = \
+            self._setup_tmt_restraint_file(REPORT_RESULT_TEARDOWN_NAME,
+                                           TMT_RESTRAINT_TEARDOWN_SCRIPT,
+                                           RSTRNT_REPORT_RESULT_SCRIPT_PATHS)
+        guest.push(self.workdir, options=DEFAULT_RSYNC_OPTIONS + ["-p"])
+        return template, setup, teardown
+
+    def _setup_tmt_restraint_file(self, filename, content, replacement=None):
+        """ Set up the TMT Restraint file. """
+        if replacement:
+            if isinstance(replacement, tuple):
+                content = \
+                    content.replace(REPLACEMENTMARKER, " ".join(replacement))
+            elif isinstance(replacement, str):
+                content = content.replace(REPLACEMENTMARKER, replacement)
+        file_path = os.path.join(self.workdir, filename)
+        with open(file_path, 'w') as template:
+            template.write(content)
+        # Make it executable, keep perms when pushing
+        perms = os.stat(file_path)
+        os.chmod(file_path, perms.st_mode | stat.S_IEXEC)
+        return file_path
+
     @contextlib.contextmanager
-    def _setup_reboot(self, guest):
-        """ Prepare the guest environment for potential reboot """
-        # Ignore local provision
-        if isinstance(guest, GuestLocal):
-            yield
-            return
-        backed_up = []
-        template, setup, teardown = None, None, None
+    def _setup_tmt_restraint(self, guest):
+        """ Prepare the guest with the TMT mocked Restraint scripts """
+        teardown_scripts = {}
+        scripts = ["report-result", "reboot"]
         try:
-            self.debug("Setup our reboot script implementations.", level=2)
-            template, setup, teardown = self._setup_reboot_script(guest)
-            try:
-                stdout = guest.execute(f"\"{setup}\" \"{template}\"")[0]
-                backed_up.extend(stdout.splitlines())
-            except tmt.utils.RunError as error:
-                if "Read-only file system" not in error.stderr:
-                    raise error
+            for script in scripts:
```

Comment:
could we refactor this to a function or something, there are too many levels here, and it is hardly readable

Reply:
I've tidied the code by moving the encapsulating try and except blocks, and debug log output, to within the method being called. This had reduced the size of the _setup_tmt_restraint() method, so I believe it is easier to read now.